patches/tModLoader/Terraria/ModLoader/ModTypeLookup.cs

```diff
@@ -0,0 +1,36 @@
+﻿using System;
+using System.Collections.Generic;
+using Terraria.Localization;
+
+namespace Terraria.ModLoader
+{
+	internal class ModTypeLookup
+	{
+		public static event Action OnClear;
+
+		public static void Clear() => OnClear?.Invoke();
+	}
+
+	internal class ModTypeLookup<T> where T : IModType
+	{
+		private static Dictionary<string, T> dict = new Dictionary<string, T>();
```

Comment:
I have some performance concerns about this has-it-all dictionary being used via Mod.Get/TryGet. Perhaps a secondary `Dictionary<Mod, Dictionary<string, T>>` would be better for those cases?

Reply:
Dict lookup is O(1) right :'), Honestly you'd have to profile the benefits of tiered lookup but I don't expect it to be more than 30% faster.

Reply:
```
Preparing 8 mods with 10000 content each.
Single Dictionary. Retrieved 80000 items in 25ms. Avg: 324ns
Tiered Dictionary. Retrieved 80000 items in 79ms. Avg: 994ns

Preparing 8 mods with 1000 content each.
Single Dictionary. Retrieved 8000 items in 2ms. Avg: 332ns
Tiered Dictionary. Retrieved 8000 items in 4ms. Avg: 605ns

Preparing 50 mods with 1000 content each.
Single Dictionary. Retrieved 50000 items in 20ms. Avg: 404ns
Tiered Dictionary. Retrieved 50000 items in 37ms. Avg: 750ns

Preparing 50 mods with 100 content each.
Single Dictionary. Retrieved 5000 items in 1ms. Avg: 394ns
Tiered Dictionary. Retrieved 5000 items in 2ms. Avg: 532ns
```

https://gist.github.com/Chicken-Bones/b1ee052214521f87b4c0ca9d4fcb1d33

Reply:
@Chicken-Bones well of course they're skewed by that. I was suggesting only using the tiered dictionary in cases when there's already separate Mod and content name parameters. I did a test like that, with non-string mod keys too just in case. Looks like Dictionary<string, Dictionary<string, T>> would be ~x2.5 faster at average.

```
Running...
Preparing 8 mods with 10000 content each.
Single Dictionary. Retrieved 80000 items in 18ms. Avg: 228ns
Tiered Dictionary (object mod keys). Retrieved 80000 items in 8ms. Avg: 110ns
Tiered Dictionary (string mod keys). Retrieved 80000 items in 7ms. Avg: 90ns

Preparing 8 mods with 1000 content each.
Single Dictionary. Retrieved 8000 items in 0ms. Avg: 95ns
Tiered Dictionary (object mod keys). Retrieved 8000 items in 0ms. Avg: 93ns
Tiered Dictionary (string mod keys). Retrieved 8000 items in 0ms. Avg: 79ns

Preparing 50 mods with 1000 content each.
Single Dictionary. Retrieved 50000 items in 11ms. Avg: 224ns
Tiered Dictionary (object mod keys). Retrieved 50000 items in 5ms. Avg: 104ns
Tiered Dictionary (string mod keys). Retrieved 50000 items in 4ms. Avg: 86ns

Preparing 50 mods with 100 content each.
Single Dictionary. Retrieved 5000 items in 0ms. Avg: 87ns
Tiered Dictionary (object mod keys). Retrieved 5000 items in 0ms. Avg: 97ns
Tiered Dictionary (string mod keys). Retrieved 5000 items in 0ms. Avg: 82ns
```

https://gist.github.com/Mirsario/ec340a2a94138dc7098656eca0997eaf

Reply:
Added that secondary dictionary in https://github.com/tModLoader/tModLoader/commit/d5431444418cc67fd6b83b49a07008b971ec03e3